### PR TITLE
Block CDN attachment scam

### PR DESCRIPTION
**Pull request by Zabuzard**
## Summary

This extends the scam blocker to also detect CDN based attachment scam. That is, people posting discord-provided image links containing scam, such as:

example

We already had similiar logic but only for actual attachments. These are "just links" in the message and not technicially attachments from Discords perspective.

The new case is also covered by unit tests and was tested locally.

## Details

The existing code for handling suspicious attachments was tweaked so it can also be used for this scenario. Therefore, we created a small wrapper for `Attachment` that can now either represent an actual Discord attachment or just an URL. The logic then uses this wrapper instead.

The URL analyse code received some rework, in particular the introduction of `AnalyseUrlResult`. This allows us to hold additional details for each URL (for example the contained attachment, if any).

These data classes received a couple of helper methods that mostly deal with string extraction (some `indexOf` and `substring` fun).

All this was done to keep the main flow of how the analyse works intact and not add a bunch of extra if-else branching just for this new edge case.

## Config

No changes are required.

The host-whitelist addition `"cdn.discordapp.com"` was already added to the TEST and PROD bots config.

### CODE REVIEW

**Review by SquidXTV — DISMISSED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            results.markNonUrlTokenFound();
+        }
+    }
+
+    private boolean containsSuspiciousKeyword(String token) {
+        String preparedToken = token.toLowerCase(Locale.US);
+
+        // Checks the token against various keywords from the config
+        // The keywords support some regex-inspired syntax
+        return config.getSuspiciousKeywords()
```
Comment (SquidXTV) [inline application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java:60]: Why arent we using actual regex here?

File: application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+     * @param results metrics representing how suspicious the token is
+     */
+    void analyze(String token, AnalyseResults results) {
+        if (token.isBlank()) {
+            return;
+        }
+
+        if (!results.pingsEveryone()
+                && ("@everyone".equalsIgnoreCase(token) || "@here".equalsIgnoreCase(token))) {
+            results.markPingsEveryone();
```
Comment (SquidXTV) [inline application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java:36]: These pre-checks really needed? They seem more like optimization, but would make the code a lot cleaner without them.

**Review by SquidXTV — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java (hunk 1 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java:60]: mh yeah... originally it wasnt needed and what i wanted to avoid is having to make a readable list of keywords ugly by adding `.*foo.*` around all of them.
originall it was just a list of keywords which the code then essentially checks with `.contains` (i.e. implicit `.*foo.*`).

now, time has passed and we needed explicit checks for "the token begins with this" or "the token ends with this".

full regex still isnt needed and i would argue that it would make the keyword list in the config uglier to read and maintain. but it would have the benefit to make the code easier to read, thats absolutely right.

like, this
```json
"keywords": [ "foo", "bar", "^baz", "boo$", "beep"]
```
i find easier to read than
```json
"keywords": [ ".*foo.*", ".*bar.*", "^baz.*", ".*boo$", ".*beep.*"]
```

although u could add as counterargument that "inventing a new syntax" is equally confusing to a maintainer. tricky situation. perhaps it should be changed.

File: application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java (hunk 2 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java:36]: u mean that `if (!flag is set && actual checks) { set flag }` could be changed to just `if (actual checks) { set flag }`.

yeah, absolutely. i think i did it bc this code actually runs on every single message send in any channel on the entirey server. and not just "when help thread is created" or similar like most other stuff.

the everyone-ping-check isnt expensive, but some of the other checks are non trivial. in particular, the keyword iteration shouldnt be executed on every single token of every single message on the entire server if we already know that there was a sus keyword.

maybe its premature optimization

File: application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java (hunk 1 above)
Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java:60]: I am fine with it for now because the "new syntax" is kinda just a subset of regex.

File: application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java (hunk 2 above)
Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java:36]: Hm I was thinking about maybe putting the check in the implementation instead, but that would require `AnalyseResult` as an Argument. For example:
```java
if (!results.containsSuspiciousKeyword() && containsSuspiciousKeyword(token)) {
            results.markContainsSuspiciousKeyword();
}
```
vs
```java
if (containsSuspiciousKeyword(results, token)) {
            results.markContainsSuspiciousKeyword();
}
```
and
```java
private boolean containsSuspiciousKeyword(AnalyseResults results, String token) {
        if (results.containsSuspiciousKeyword()) {
                return true;
        }
        // actual impl
        String preparedToken = token.toLowerCase(Locale.US);
        ...
}
```

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java:36]: great idea. has the risk to confuse in regards to who is responsible for also setting the flag though. but that would be a heavy side effect for a method called `contains...`, so it should be okay.

i like the idea, lets do it like that :)

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/moderation/scam/TokenAnalyse.java:36]: (ran into a logic issue doing that, which turned out to be less readable/preferable. we decided to keep it as-is for now)
